### PR TITLE
Routes

### PR DESCRIPTION
**Pull request by danhanson**
for more information, view the websocket API documentation

Here is how to get a vehicle's route:

send this through the socket

```
{ "route":{"model":"Vehicle","id":3}}
```

and receive:

```
{ "routed":{
    "model": "Vehicle",
    "value": {
      "vehicle":3,
      "actions":[
        { "action":"pickup","latitude":123,"longitude":456, "commodity":7 },
        { "action":"dropoff","latitude":789,"longitude":777, "commodity":7}
      ]
    }
  }
}
```

### CODE REVIEW

**Comment by aj-michael** [comment]
Dan can you add some logging for the routes? Specifically for the error cases, right now there is no record on the server of what went wrong.

File: app/io/pathfinder/models/Vehicle.scala [hunk 1]
```diff
@@ -66,4 +66,7 @@ class Vehicle() extends Model {
 
     @Column(nullable=false)
     var capacity: Int = 0
+
+    @ManyToOne
```
Comment (aj-michael) [inline app/io/pathfinder/models/Vehicle.scala]: I believe we decided against this @danhanson

File: app/io/pathfinder/routing/Action.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def longitude: Double
+    def commodity: Long
+}
+
+object Action {
+
+    case class PickUp(
+        latitude: Double,
+        longitude: Double,
+        commodity: Long
```
Comment (aj-michael) [inline app/io/pathfinder/routing/Action.scala]: Should this be commodityId?

File: app/io/pathfinder/routing/Action.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            com.startLatitude,
+            com.startLongitude,
+            com.id
+        )
+    }
+
+    case class DropOff(
+        latitude: Double,
+        longitude: Double,
+        commodity: Long
```
Comment (aj-michael) [inline app/io/pathfinder/routing/Action.scala]: Same as above.

File: app/io/pathfinder/routing/Action.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import io.pathfinder.models.Commodity
+import play.api.libs.json.{Writes, Json, JsPath, Format}
+
+sealed abstract class Action(val name: String) {
+    def latitude: Double
+    def longitude: Double
+    def commodity: Long
+}
+
+object Action {
```
Comment (aj-michael) [inline app/io/pathfinder/routing/Action.scala:11]: Do you think we should have timestamps on the actions?

File: app/io/pathfinder/websockets/ModelTypes.scala [hunk 5]
```diff
@@ -6,15 +6,11 @@ import play.api.libs.json._
  * An enum containing the models that a websocket message can use
  */
 object ModelTypes extends Enumeration {
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/ModelTypes.scala:8]: Why did you get rid of reads and writes? I though that was the whole point of this class.

Please add comments explaining the "new" point of the class.

File: app/io/pathfinder/websockets/WebSocketMessage.scala [hunk 6]
```diff
@@ -77,16 +77,30 @@ object WebSocketMessage {
...
   implicit val deleteFormat = Json.format[Delete]
 
+
+  case class Route(
+    model: ModelType,
+    id:    Long
+  ) extends ControllerMessage
+  implicit val routeFormat = Json.format[Route]
+
+  case class Routed(
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/WebSocketMessage.scala:96]: What is a Routed and how is it different from a Route? Please add comments explaining.

File: app/io/pathfinder/websockets/controllers/VehicleSocketController.scala [hunk 7]
```diff
@@ -1,9 +1,29 @@
...
+import io.pathfinder.routing.Action.{DropOff, PickUp}
+import io.pathfinder.routing.{Action, Route}
+import io.pathfinder.websockets.{WebSocketMessage, ModelTypes}
+import io.pathfinder.websockets.WebSocketMessage.{Route => RouteMsg, Routed}
+import scala.collection.JavaConversions.asScalaBuffer
 /**
  * manages vehicle API calls
  */
-object VehicleSocketController extends WebSocketCrudController[Vehicle](ModelTypes.Vehicle,Vehicle.Dao)
+object VehicleSocketController extends WebSocketCrudController[Vehicle](ModelTypes.Vehicle,Vehicle.Dao) {
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/controllers/VehicleSocketController.scala:12]: Why do we need to specify the Vehicle.Dao argument separately? Can't we infer that from the ModelType?

File: app/io/pathfinder/websockets/controllers/WebSocketCrudController.scala [hunk 8]
```diff
@@ -3,8 +3,8 @@ package io.pathfinder.websockets.controllers
 import io.pathfinder.data.{CrudDao, Resource}
 import io.pathfinder.websockets.ModelTypes.ModelType
 import io.pathfinder.websockets.WebSocketMessage
-import io.pathfinder.websockets.WebSocketMessage.{Create, Created, Delete, Deleted, Model, Read, Update, Updated}
-import play.api.libs.json.{JsResult, Reads, Writes}
+import io.pathfinder.websockets.WebSocketMessage._
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/controllers/WebSocketCrudController.scala:6]: Please turn off your IntelliJ setting that is collapsing imports.

File: app/io/pathfinder/websockets/controllers/WebSocketCrudController.scala [hunk 9]
```diff
@@ -24,10 +24,10 @@ abstract class WebSocketCrudController[V <: ebean.Model](
         Logger.info(s"Received web socket crud request $webSocketMessage")
         webSocketMessage match {
             case Update(m, id, value) => Some(Updated(model, writes.writes(dao.update(id, resources.reads(value).get).get)))
-            case Create(m, value) => { Some(Created(model, writes.writes(dao.create(resources.reads(value).get).get))) }
+            case Create(m, value) => Some(Created(model, writes.writes(dao.create(resources.reads(value).get).get)))
             case Delete(m, id) => Some(Deleted(model, writes.writes(dao.delete(id).get)))
             case Read(m, id) => Some(Model(model, writes.writes(dao.read(id).get)))
-            case x => None
+            case x: WebSocketMessage => Some(Error("No support for message: " + WebSocketMessage.format.writes(x) + " for model: "+model.toString))
```
Comment (aj-michael) [inline app/io/pathfinder/websockets/controllers/WebSocketCrudController.scala:50]: This never gets used: The server errors on a None.get and returns nothing. Please fix ASAP. See #19

**Comment by aj-michael** [comment]
Also Dan, please add the routes socket API to the Wiki page.

File: app/io/pathfinder/routing/Action.scala (hunk 4 above)
Comment (danhanson) [inline app/io/pathfinder/routing/Action.scala:11]: What would the timestamp represent? Would it be the time we expect the action to occur?

**Comment by aj-michael** [comment]
Dan, can you change it so that the first action in the routes list is the vehicles current location? The route drawn on the iOS map doesn't make sense to start at the first pickup, it should start where the vehicle currently is.

I opened an issue for this: #21

File: app/io/pathfinder/websockets/controllers/VehicleSocketController.scala (hunk 7 above)
Comment (danhanson) [inline app/io/pathfinder/websockets/controllers/VehicleSocketController.scala:12]: I guess you could. I made the modeltypes mostly for validating/parsing and multiplexing requests. I didn't consider the daos at the time I made it.

**Comment by aj-michael** [comment]
The branch is at a stable point. More tests will be added next sprint, sorry for the lack of code coverage; they will be added ASAP next sprint.

See #24 for the continuation of this work that involves pull requests.

**Pull request merged**